src/modules/kvs/lookup.c

```diff
@@ -696,6 +696,135 @@ int lookup_set_aux_data (lookup_t *lh, void *data)
     return -1;
 }
 
+/* return 0 on success, -1 on failure.  On success, stall should be
+ * checked */
+static int get_single_blobref_valref_value (lookup_t *lh, bool *stall)
+{
+    struct cache_entry *hp;
+    const char *reftmp;
+    void *valdata;
+    int len;
+
+    if (!(reftmp = treeobj_get_blobref (lh->wdirent, 0))) {
+        lh->errnum = errno;
+        return -1;
+    }
+    if (!(hp = cache_lookup (lh->cache, reftmp, lh->current_epoch))
+        || !cache_entry_get_valid (hp)) {
+        set_single_missing_ref (lh, reftmp, true);
+        (*stall) = true;
+        return 0;
+    }
+    if (!(valdata = cache_entry_get_raw (hp, &len))) {
+        flux_log (lh->h, LOG_ERR, "valref points to non-raw data");
+        lh->errnum = EPERM;
+        return -1;
+    }
+    if (!(lh->val = treeobj_create_val (valdata, len))) {
+        lh->errnum = errno;
+        return -1;
+    }
+    (*stall) = false;
+    return 0;
+}
+
+/* return 0 on success, -1 on failure.  On success, stall should be
+ * check */
+static int get_multi_blobref_valref_value (lookup_t *lh, int refcount,
```

Comment:
Just a suggestion and may just be a style thing:  would it be clearer if determining the aggregate size and copying to the aggregate buffer were split into two functions?

Reply:
yeah, that could probably help.